### PR TITLE
state: Multple interfaces can be added to machines; more tests/methods.

### PR DESCRIPTION
**Pull request by dimitern**
Replaced singular Machine.AddInterface() by bulk Machine.AddInterfaces() method, which adds
multiple interfaces to the same machine, in a single transaction. Test coverage now is as good as
it can reasonably get, without resorting to ugly patching. Added jujutxn.TestHook tests and a speculative highly concurrent test for AddInterfaces.

New methods added and tested for all corner cases:
- Machine.AllInterfaces()
- Machine.RemoveAllInterfaces()
- Interface.Remove()
- Interface.Machine()

Defined two new concrete error types:
- ErrProviderIDNotUnique, which will be used for Spaces and Subnets later.
- ErrParentInterfaceHasChildren, returned when trying to remove an interface with children.  

Still left to do in follow-ups:
- Interface links state model - one or more links per interface, each having a unique ID and usable instead of a "naked" address across the board; 
- Machine.UpdateInterfaces() taking variable InterfaceArgs structs like AddInterfaces();
- Removing interfaces on machine destruction;
- Adding interfaces during at provisioning time;
- Updating / adding interfaces to match what's observed by MA at startup;
- more..

### CODE REVIEW

File: state/interfaces_test.go [hunk 1]
```diff
@@ -302,3 +399,361 @@ func (s *interfacesStateSuite) TestMachineInterfaceReturnsInterface(c *gc.C) {
...
+
+func (s *interfacesStateSuite) TestMachineRemoveAllInterfacesNoErrorIfNoInterfacesExist(c *gc.C) {
+	s.assertNoInterfacesOnMachine(c, s.machine)
+
+	err := s.machine.RemoveAllInterfaces()
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *interfacesStateSuite) TestInterfaceRefreshUpdatesStaleDocData(c *gc.C) {
+	fooInterface := s.addSimpleInterface(c)
```
Comment (dooferlad) [inline state/interfaces_test.go]: This test is unclear. Please add a comment telling us how fooInterface.Refresh() is able to change hardware address. It seems that fooInterface.Refresh is looking up some details by interface name.

File: state/interfaces_test.go [hunk 2]
```diff
@@ -302,3 +399,361 @@ func (s *interfacesStateSuite) TestMachineInterfaceReturnsInterface(c *gc.C) {
...
+			err := s.machine.AddInterfaces(testArgs...)
+			c.Logf("testing %q -> %v", testAbout, err)
+			if err != nil {
+				c.Assert(err, jc.Satisfies, isAlreadyExistsOrProviderIDNotUniqueError)
+			} else {
+				select {
+				case successfulArgs <- testArgs:
+				default:
+					// successfulArgs is buffered, so if we can't send there was
+					// more than once success.
```
Comment (dooferlad) [inline state/interfaces_test.go]: s/once/one/

File: state/interfaces_test.go [hunk 3]
```diff
@@ -302,3 +399,361 @@ func (s *interfacesStateSuite) TestMachineInterfaceReturnsInterface(c *gc.C) {
...
+					c.Fatalf("unexpected more than one success for args %+v", testArgs)
```
Comment (dooferlad) [inline state/interfaces_test.go]: unexpected: more than one...

**Comment by dimitern** [comment]
Thanks for the detailed reviews! I'll start replying shortly in order.

File: state/interfaces_test.go (hunk 1 above)
Comment (dimitern) [inline state/interfaces_test.go]: Indeed, now looking at it seems so. What it's trying to test is 2 copies of state.Interface with the same docID can hold different field values and Refresh() updates the doc and values. Unfortunately, there's no way to change anything of an interface - only add and remove. Other similar tests in state do things like: get original, copy with the same docID, copy.EnsureDead(), check original.Life() == Alive && copy.Life() == Dead, original.Refresh() && original.Life() == Dead. So this test "cheats" by removing the doc and adding a new one with the same docID but set hardware address and relies on Refresh() to get the new doc by ID. Too many assumptions, will simplify or just drop perhaps.

File: state/interfaces_test.go (hunk 2 above)
Comment (dimitern) [inline state/interfaces_test.go]: Will fix.

File: state/interfaces_test.go (hunk 3 above)
Comment (dimitern) [inline state/interfaces_test.go]: Will do.

**Comment by dimitern** [comment]
@frobware @dooferlad Updated as described and rebased onto maas-spaces2 tip, please have another look and see if it's good to land.

**Comment by dooferlad** [comment]
LGTM

**Comment by dimitern** [comment]
$$merge$$

**Pull request merged**